File: Tools/1 - Common/MoCommon/IO/RDirectory.cs
using System.Diagnostics;
using System.IO;
using MO.Common.Lang;

namespace MO.Common.IO
{
   //============================================================
   // <T>目录工具类。</T>
   //============================================================
   public static class RDirectory
   {
      //============================================================
      // <T>判断目录是否存在。</T>
      //============================================================
      public static bool Exists(string directory) {
         return Directory.Exists(directory);
      }

      //============================================================
      // <T>判断指定文件是否存在。</T>
      //
      // @param 文件地址
      // @return 结果
      //============================================================
      public static bool IsFileExists(string directory) {
         int dirIndex = directory.LastIndexOf("\\");
         string dirPath = directory.Substring(0, directory.Length - (directory.Length - dirIndex - 1));
         if (Directory.Exists(dirPath)) {
            string[] filePaths = Directory.GetFiles(dirPath);
            foreach (string path in filePaths) {
               if (directory.Equals(path)) {
                  return true;
               }
            }
         }
         return false;
      }

      //============================================================
      // <T>获得文件夹下文件列表。</T>
      //
      // @param paths 文件列表
      // @param directory 文件夹
      // @param deep 深度
      //============================================================
      private static bool InnerListFiles(FStrings paths, string directory, bool deep, string extension, bool constainsDir, bool constainsFile) {
         if(!Directory.Exists(directory)){
            return false;
         }
         // 搜索子文件夹
         foreach (string dir in Directory.GetDirectories(directory)) {
            if (constainsDir) {
               paths.Push(dir);
            }
            if (deep) {
               InnerListFiles(paths, dir, deep, extension, constainsDir, constainsFile);
            }
         }
         // 搜索文件列表
         if (constainsFile) {
            foreach (string file in Directory.GetFiles(directory)) {
               if (null == extension) {
                  paths.Push(file);
               } else if (file.EndsWith(extension)) {
                  paths.Push(file);
               }
            }
         }
         return true;
      }

      //============================================================
      // <T>列出文件夹下所有文件。</T>
      //
      // @param directory 文件夹
      // @param deep 深度
      //============================================================
      public static FStrings ListFiles(string directory, bool deep = false, string extension = null) {
         FStrings files = new FStrings();
         InnerListFiles(files, directory, deep, extension, false, true);
         files.Sort();
         return files;
      }

      //============================================================
      // <T>列出文件夹下所有子文件夹。</T>
      //
      // @param directory 文件夹
      // @param deep 深度
      //============================================================
      public static FStrings ListDirectories(string directory, bool deep = false) {
         FStrings directories = new FStrings();
         InnerListFiles(directories, directory, deep, null, true, false);
         directories.Sort();
         return directories;
      }

      //============================================================
      // <T>如果目录中文件夹不存在则创建文件夹。</T>
      //
      // @param directory 文件夹目录
      //============================================================
      public static void MakeDirectories(string directory) {
         if (!Directory.Exists(directory)) {
            Directory.CreateDirectory(directory);
         }
      }

      //============================================================
      // <T>如果目录中文件夹不存在则创建文件夹。</T>
      //
      // @param directory 文件夹目录
      //============================================================
      public static void MakeDirectoriesForFile(string fileName) {
         fileName = fileName.Replace('/', '\\');
         int index = fileName.LastIndexOf('\\');
         if (-1 != index) {
            string directory = fileName.Substring(0, index);
            if (!Directory.Exists(directory)) {
               Directory.CreateDirectory(directory);
            }
         }
      }

      //============================================================
      // <T>打开指定路径。</T>
      //============================================================
      public static void ProcessOpen(string path) {
         Process.Start(new ProcessStartInfo(path));
      }

      //============================================================
      // <T>复制来源路径所有文件到目标路径。</T>
      //
      // @param sourcePath 来源路径
      // @param targetPath 目标路径
      // @param includes 包含列表
      // @param excludes 排除列表
      //============================================================
      public static void Copy(string sourcePath, string targetPath, FStrings includes, FStrings excludes) {
      }

      //============================================================
      // <T>清空指定目录内所有文件。</T>
      //
      // @param path 指定目录
      //============================================================
      public static void Clear(string path) {
         if (Directory.Exists(path)) {
            string[] fileNames = Directory.GetFiles(path);
            foreach (string fileName in fileNames) {
               File.Delete(fileName);
            }
         }
      }

      //============================================================
      // <T>删除指定目录。</T>
      //
      // @param path 指定目录
      //============================================================
      public static void Delete(string path) {
         if (Directory.Exists(path)) {
            Directory.Delete(path, true);
         }
      }
   }
}

File: Platform/mo-1-common/src/lang-java/org/mo/com/io/RDirectory.java
package org.mo.com.io;

import java.io.File;
import org.mo.com.lang.FStrings;

//============================================================
// <T>目录工具类。</T>
//============================================================
public class RDirectory
{
   //============================================================
   // <T>列出目录中所有文件。</T>
   //
   // @param directory 目录名称
   // @param directories 目录集合
   // @param files 文件集合
   // @param deep 递归
   // @param startsWith 开始字符串
   // @param endsWith 结束字符串
   //============================================================
   protected static void innerListDirectory(File directory,
                                            FFileInfos infos,
                                            boolean deep,
                                            String startsWith,
                                            String endsWith){
      if(directory.isDirectory()){
         for(File file : directory.listFiles()){
            if(deep && file.isDirectory()){
               innerListDirectory(file, infos, deep, startsWith, endsWith);
            }else{
               // 选择过滤器
               String fileName = file.getName();
               boolean finded = false;
               if((null == startsWith) && (null == endsWith)){
                  finded = true;
               }else if((null != startsWith) && (null == endsWith)){
                  if(fileName.startsWith(startsWith)){
                     finded = true;
                  }
               }else if((null == startsWith) && (null != endsWith)){
                  if(fileName.endsWith(endsWith)){
                     finded = true;
                  }
               }else if((null != startsWith) && (null != endsWith)){
                  if(fileName.startsWith(startsWith) && fileName.startsWith(endsWith)){
                     finded = true;
                  }
               }
               if(finded){
                  // 增加一个文件
                  FFileInfo info = new FFileInfo();
                  info.setFileName(file.getAbsolutePath());
                  infos.push(info);
               }
            }
         }
      }
   }

   //============================================================
   // <T>列出目录中所有文件。</T>
   //
   // @param directory 目录名称
   // @return 文件信息集合
   //============================================================
   public static FFileInfos listFiles(String directory){
      FFileInfos infos = new FFileInfos();
      innerListDirectory(new File(directory), infos, true, null, null);
      return infos;
   }

   //============================================================
   // <T>列出目录中所有文件。</T>
   //
   // @param directory 目录名称
   // @param deep 递归
   // @return 文件信息集合
   //============================================================
   public static FFileInfos listFiles(String directory,
                                      boolean deep){
      FFileInfos infos = new FFileInfos();
      innerListDirectory(new File(directory), infos, deep, null, null);
      return infos;
   }

   //============================================================
   // <T>列出目录中所有文件。</T>
   //
   // @param directory 目录名称
   // @param startsWith 开始字符串
   // @param endsWith 结束字符串
   // @return 文件信息集合
   //============================================================
   public static FFileInfos listFiles(String directory,
                                      String startsWith,
                                      String endsWith){
      FFileInfos infos = new FFileInfos();
      innerListDirectory(new File(directory), infos, true, startsWith, endsWith);
      return infos;
   }

   /**
    * <p>列出指定目录下的所有目录名称列表</p>
    * 
    * @param directory 指定目录
    * @return 目录名称列表
    */
   public static FStrings listDirectory(String directory){
      return listDirectory(directory, false);
   }

   /**
    * <p>列出指定目录下的所有目录名称列表</p>
    * 
    * @param directory 指定目录
    * @return 目录名称列表
    */
   public static FStrings listDirectory(String directory,
                                        boolean fullName){
      FStrings list = new FStrings();
      File dir = new File(directory);
      if(dir.isDirectory()){
         for(File file : dir.listFiles()){
            if(file.isDirectory()){
               if(fullName){
                  list.push(file.getAbsolutePath());
               }else{
                  list.push(file.getName());
               }
            }
         }
      }
      return list;
   }
}
